src/subnet.c

```diff
@@ -37,35 +37,86 @@
 /* lists type of subnet */
 
 splay_tree_t *subnet_tree;
+static uint32_t hash_seed;
 
 /* Subnet lookup cache */
 
-static hash_t *ipv4_cache;
-static hash_t *ipv6_cache;
-static hash_t *mac_cache;
+static uint32_t hash_function_ipv4_t(const ipv4_t *p) {
+	/*
+	This basic hash works because
+	a) Most IPv4 networks routed via tinc are not /0
+	b) Most IPv4 networks have more unique low order bits
+	*/
+	uint16_t *halfwidth = (uint16_t *)p;
+	// 10.0.x.x/16 part
+	uint32_t hash = hash_seed;
+	hash += halfwidth[1] * 0x9e370001UL;
+	// x.x.0.[0-255] part
+
+#if _____LP64_____
+	return hash ^ halfwidth[0];
+#else
+	// ensure that we have a /24 with no collisions on 32bit
+	return hash ^ ntohs(halfwidth[0]);
```

Comment:
This looks wrong. There is nothing in the hash table code itself which depends on the pointer size. So what really is going on is that the size of the hash table is what decides which variant is better. But even then there is a problem: on big-endian architectures, `ntohs()` is a no-op, and the order of `halfwidth` would be reversed as well. So I would really like to see something without this magic. Accessing the individual bytes using `p->x[0..3]` would be better. You could perhaps use the first three octets of the IP address and multiply them by the prime, and then XOR the result with the last octet.

Reply:
Endian safety introduced.

 A `uint8_t*` array feels unnecessary. A 16bit rotation should be more than enough.